plutus-core/plutus-ir/src/PlutusIR/Transform/CaseOfCase.hs

```diff
@@ -0,0 +1,198 @@
+caseOfCase binfo conappsOnly t0 = do
+    -- We are going to record information about variables in a global map, so we
+    -- need global uniqueness
+    t <- PLC.rename t0
+    let vinfo = termVarInfo t
+    pure $ transformOf termSubterms (processTerm binfo vinfo conappsOnly) t
+
+processTerm ::
+    forall tyname name uni fun a .
+    (Ord fun, PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique) =>
+    BuiltinsInfo uni fun ->
+    VarsInfo tyname name ->
+    Bool ->
+    Term tyname name uni fun a ->
+    Term tyname name uni fun a
+processTerm binfo vinfo conappsOnly t
+    -- We have a saturated datatype matcher application
+    | Just (SplitMatchContext outerVars (outerScrut, outerScrutAnn) (outerResTy, outerResTyAnn) outerBranches, _, reconstructOuter) <- splitMatch binfo vinfo t
+
+    -- The scrutinee is itself an application
+    -- Note: we don't use the inner result type, we're going to replace it
+    , Just (SplitMatchContext innerVars (innerScrut, innerScrutAnn) _ innerBranches, innerBranchArities, reconstructInner) <- splitMatch binfo vinfo outerScrut
+    , let
+        mkNewOuterMatch newScrut =
+          reconstructOuter $ SplitMatchContext outerVars (newScrut, outerScrutAnn) (outerResTy, outerResTyAnn) outerBranches
+
+    -- See Note [Case-of-case and conapps]
+    , Just newInnerBranches <- mapBranches (\b -> if not conappsOnly || isConApp b then Just (mkNewOuterMatch b) else Nothing) innerBranches innerBranchArities
+    , let
+        newInnerMatch =
+          reconstructInner $ SplitMatchContext innerVars (innerScrut, innerScrutAnn) (outerResTy, outerResTyAnn) newInnerBranches
+
+      = newInnerMatch
+    | otherwise = t
+    where
+      -- We know this value gets cased on, so if it's a constructor application it _must_ be saturated:
+      -- if it was undersaturated the result would be a function (which can't be cased on), and if it was
+      -- oversaturated then it wouldn't typecheck since you can't apply a datatype value
+      isConApp (splitApplication -> (Var _ n, _)) | Just (DatatypeConstructor{}) <- lookupVarInfo n vinfo = True
+      isConApp _ = False
+
+-- | Apply the given function to the term "inside" the case branches in the given 'AppContext'.
+-- Must be given an arity for each branch so it knows how many binders to go under.
+mapBranches ::
+  forall tyname name uni fun a
+  . (Term tyname name uni fun a -> Maybe (Term tyname name uni fun a))
+  -> AppContext tyname name uni fun a
+  -> [Arity]
+  -> Maybe (AppContext tyname name uni fun a)
+mapBranches f = go
+    where
+    go :: AppContext tyname name uni fun a -> [Arity] -> Maybe (AppContext tyname name uni fun a)
+    go AppContextEnd [] = Just AppContextEnd
+    go (TermAppContext branch ann ctx) (arity:arities) =
+      TermAppContext <$> underBindersMaybe arity f branch <*> pure ann <*> go ctx arities
+    go _ _ = Nothing
+
+    -- I tried to come up with a way to do this in terms of the underBinders traversal
```

Comment:
I don't get it. underbinders should just work since its a traversal with Maybe as the Applicative?

```
    go (TermAppContext branch ann ctx) (arity:arities) =
      TermAppContext <$> underBinders arity f branch <*> pure ann <*> go ctx arities
```